### PR TITLE
planner: fix the issue that some IndexJoin cannot use plan-cache

### LINKED ISSUE #29101: pr#28915 results in performance degradation in TPCC benchmark
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

PR https://github.com/pingcap/tidb/pull/28915 results in performance degradation in TPCC benchmark. The performance of TPCC drops 5.4%-10%.

1. create a new tidb cluster  
2. set prepared-plan-cache.enabled: true
3. run TPC-C 1000 warehouse benchmark.
 ```
go-tpc -H $tidb -P9999 -D tpcc tpcc run --time 10m --warehouses 1000 --threads 64
```

### 3. What did you see instead (Required)
The following statement cannot use plan cache. 
```
| 39601b01e7dd65eca522e2c269a675159880ec0925e552ede4658fa9c077838d | SELECT c_discount, c_last, c_credit, w_tax FROM customer, warehouse WHERE w_id = ? AND c_w_id = w_id AND c_d_id = ? AND c_id = ? [arguments: (936, 7, 158)];   |
| c360a20a84bf1bf59dad0277e145673a469241273114901c3a752a9ff1543687 | SELECT /*+ TIDB_INLJ(order_line,stock) */ COUNT(DISTINCT (s_i_id)) stock_count FROM order_line, stock  WHERE ol_w_id = ? AND ol_d_id = ? AND ol_o_id < ? AND ol_o_id >= ? - 20 AND s_w_id = ? AND s_i_id = ol_i_id AND s_quantity < ? [arguments: (391, 1, 3058, 3058, 391, 18)];  |
```
### 4. What is your TiDB version? (Required)

| Release Version: v5.3.0-alpha-1136-gae346a553
Edition: Community
Git Commit Hash: ae346a553088c56dc8c17f0e1a50c4008be59bc8
Git Branch: master
UTC Build Time: 2021-10-18 16:49:13
GoVersion: go1.16.4
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: close #29101

Problem Summary: planner: fix the issue that some IndexJoin cannot use plan-cache

### What is changed and how it works?

Introduce a new interface `MutableRanges` to represent these ranges that may change after they are created.

`MutableRranges` are mainly designed for `PlanCache`, since some ranges in a cached plan have to be rebuilt when reusing.

### Check List

Tests 

- [x] Unit test

### Release note

```release-note
None
```

### CODE REVIEW

**Review by rebelice — APPROVED** [review]

File: planner/core/exhaust_physical_plans.go [hunk 1]
```diff
@@ -414,12 +414,16 @@ func (p *LogicalJoin) constructIndexJoin(
...
 	outerIdx int,
 	innerTask task,
-	ranges []*ranger.Range,
+	ranges ranger.MutableRanges,
 	keyOff2IdxOff []int,
 	path *util.AccessPath,
 	compareFilters *ColWithCmpFuncManager,
 	extractOtherEQ bool,
 ) []PhysicalPlan {
+	if ranges == nil {
```
[the remark this answers is not in the document]
Comment (qw4990) [reply planner/core/exhaust_physical_plans.go:423]: It's defensive code. Actually, I'm not sure whether it can happen, just adding it for safety.

**Pull request merged**